### PR TITLE
refactor `Lock` for parallel compiler

### PR DESCRIPTION
**Pull request by SparrowLii**
part of https://github.com/rust-lang/rust/pull/101566 
This PR refactor `Lock` for parallel compiler, facilitating code review and perf test.

ps. refactored `Lock` is not `Send` or `Sync`. It depends on https://github.com/rust-lang/rust/pull/107586 to get thread safety.

### CODE REVIEW

File: compiler/rustc_data_structures/src/sync.rs [hunk 1]
```diff
@@ -381,55 +385,106 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
     }
 }
 
-#[derive(Debug)]
-pub struct Lock<T>(InnerLock<T>);
+pub struct Lock<T> {
+    single_thread: bool,
+    data: UnsafeCell<T>,
+    borrow: Cell<bool>,
+    mutex: RawMutex,
```
Comment (joboet) [inline compiler/rustc_data_structures/src/sync.rs:392]: Couldn't these two fields be merged into a single `AtomicBool`? The single-threaded version would become `borrow.as_ptr().replace(true)`, while the multi-threaded version would use `parking_lot_core::park` to handle waiting.

File: compiler/rustc_data_structures/src/sync.rs [hunk 2]
```diff
@@ -381,55 +385,106 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
     }
 }
 
-#[derive(Debug)]
-pub struct Lock<T>(InnerLock<T>);
+pub struct Lock<T> {
+    single_thread: bool,
```
Comment (joboet) [inline compiler/rustc_data_structures/src/sync.rs:389]: Shouldn't loading `PARALLEL` with relaxed ordering (or maybe a non-atomic read) be just as efficient? That avoids having to copy around the `bool` in every lock.

File: compiler/rustc_data_structures/src/sync.rs (hunk 1 above)
Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:392]: Thanks! It makes sense, I'll try it soon

**Comment by Dylan-DPC** [comment]
r? @cjgillot 

assigning them as they are the assigned reviewer on the other pr as well

**Comment by Dylan-DPC** [comment]
@SparrowLii any updates on this?

**Comment by SparrowLii** [comment]
@Zoxc has a more completed implementation in #111713. I will first do the work of query cache next. If they still have no time to update, I will finish this PR instead.

**Comment by Dylan-DPC** [comment]
Closing this as #111713 is now merged

**Pull request closed without merging**